### PR TITLE
Make authorized_key preserve key order

### LINKED ISSUE #4780: "authorized_key" re-orders the keys and comments in the pre-existing authorized_keys file
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

system/authorized_key
##### ANSIBLE VERSION

```
ansible 2.1.1.0
```
##### OS / ENVIRONMENT

Managing from: LinuxMint
Managing: CentOS 6
##### SUMMARY

authorized_key is re-ordering the lines in a pre-existing authorized_keys file. This causes comments and the related lines to be re-ordered.
##### STEPS TO REPRODUCE

The managed host (B) should have a authorized_keys file containing multiple keys and comments. Example could look like this:

```
# The following is allice's key
ssh-rsa <key content-removed> ssh-key-comment-allice
# The following is bob's key
ssh-rsa <key content-removed> ssh-key-comment-bob
# The following is eve's key
ssh-rsa <key content-removed> ssh-key-comment-eve

```

After adding a key from another host (A) to the host (B) using the "authorized_key" task, below the pre-existing authorized_keys file as shown above is completely reordered.

```

---
- name: Generate ssh key on host (A)
  user: name=root generate_ssh_key=yes ssh_key_bits=4096 ssh_key_file=.ssh/id_rsa ssh_key_type=rsa

- name: Download generated ssh key from host (A)
  fetch: src=/root/.ssh/id_rsa.pub dest=/tmp/ansible/{{ ansible_hostname }}_id_rsa.pub 
         flat=yes validate_checksum=yes

- name: Upload/install the ssh-key from host (A) to host (B)
  authorized_key: user=root state=present 
      key="{{ lookup('file', '/tmp/ansible/{{ ansible_hostname }}_id_rsa.pub') }}"
  delegate_to: "{{ host_B }}"
```

##### EXPECTED RESULTS

```
# The following is allice's key
ssh-rsa <key content-removed> ssh-key-comment-allice
# The following is bob's key
ssh-rsa <key content-removed> ssh-key-comment-bob
# The following is eve's key
ssh-rsa <key content-removed> ssh-key-comment-eve
ssh-rsa <key content-removed> ssh-key-comment-joe
```

As shown in the example, it would be expected to see the added key (in the example "joe") at the end of the file keeping the pre-existing lines unchanged.
##### ACTUAL RESULTS

The result of the reordering might look like this containing the added key for "joe". The exact reason and ordering is not known at this point as of such the following example is just an example.

```
# The following is allice's key
ssh-rsa <key content-removed> ssh-key-comment-eve
# The following is eve's key
ssh-rsa <key content-removed> ssh-key-comment-joe
# The following is bob's key
ssh-rsa <key content-removed> ssh-key-comment-allice
ssh-rsa <key content-removed> ssh-key-comment-bob

```

### PR DESCRIPTION
**Pull request by alikins**
##### ISSUE TYPE

- Bugfix Pull Request
##### COMPONENT NAME

system/authorized_key.py
##### ANSIBLE VERSION

```
ansible 2.3.0 (authorized_key_test 3a02cded91) last updated 2016/10/20 17:18:25 (GMT -400)
  lib/ansible/modules/core: (detached HEAD d77027f495) last updated 2016/10/20 15:01:00 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 77ac192f3b) last updated 2016/10/20 15:01:00 (GMT -400)
  config file = /home/adrian/ansible/ansible.cfg
  configured module search path = ['/home/adrian/src/ansible-modules-core', '/home/adrian/src/ansible-modules-extras']


```
##### SUMMARY

Track the ordering of keys in the original file (rank)
and try to preserve it when writing out updates.

Fixes #4780

There are some related integration tests for ansible/ansible coming as well...

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @alikins for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by alikins** [comment]
https://github.com/ansible/ansible/pull/18130  for the aforementioned ansible/ansible integration tests

**Comment by abadger** [comment]
Looks good.  Made a comment on slack about not preserving comments in the key parameter but no one's asked for that yet so we can wait and see if we have any feature requests for it.  Also talked about adding a test case for existing comments inside of the file.  That's the only thing I see missing.  So feel free to merge.  Check with jimi-c about whether to cherry-pick this into the stable-2.2 branch (for 2.2.0) or temp-post-2.2.0 branch for inclusion in 2.2.1.

**Comment by gregdek** [comment]
Thanks @alikins for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by alikins** [comment]
updated to preserve comment ordering (and to preserve comments, which was busted in this pr before)

**Comment by abadger** [comment]
+1 to merge

File: system/authorized_key.py [hunk 1]
```diff
@@ -342,6 +350,7 @@ def parsekey(module, raw_key):
     # check for options
     if type_index is None:
         return None
+        #return (raw_key, None, None, None, rank)
```
Comment (abadger) [inline system/authorized_key.py]: You can remove this, since it looks like you now deal with it earlier.

**Pull request merged**